File: analysis.py
from special_items import chart_dir
import numpy as np
import pandas as pd
import scipy.stats as stats
import matplotlib.pyplot as plt
import os

def group_consecutive(a,step=1):
    ''' group consecutive numbers in an array
        modified from https://zhuanlan.zhihu.com/p/29558169'''
    return np.split(a, np.where(np.diff(a) > step)[0] + 1)

def countWithRange(target, countRange=None, name='power'):
    ''' 统计DataFrame中不同的值出现的次数
        target :: DataFrame() :: 数据源；
        countRange :: tuple(start,end) :: 节选数据的index，默认为None；
        name :: str() :: column名称，默认为\'power\'。'''
    if countRange is None:
        target_data = target[name].values
    else:
        target_data = target[name].values[countRange[0]:countRange[1]]
    utarget,uind=np.unique(target_data,return_inverse=True)
    target_count=np.bincount(uind)
    auc = target_count.sum()

    return utarget, target_count, auc

def horizontal_thresh_method(time_array, power_array, on_thresh):
    spike = np.sign(power_array-on_thresh)
    data = pd.DataFrame(np.hstack((time_array[:,np.newaxis], power_array[:,np.newaxis], spike[:,np.newaxis])),
                        columns=['time','power','spike'])
    return data

def horizontal_log_thresh_method(ch_num, time_array, power_array):
    ''' 目前计算阈值的方法：
        对power的数据取log10，用标准正态分布函数拟合[参考维基百科正态分布词条]，
        得到mu(loc)和sigma(scale)。
        阈值取10^(mu+sigma)'''

    logData = np.log10(power_array)
    loc, scale = stats.norm.fit(logData)
    on_thresh = 10**(loc+scale)

    dataSheet = horizontal_thresh_method(time_array, power_array, on_thresh)
    distribution_plot(ch_num, dataSheet)

    return dataSheet, on_thresh

def eye_data_episode(eye_data_sheet, episode_gap=4):
    ''' 对视频分析的结果进行episode归类
        返回值：list(tuple(start,end),...)'''
    raw_data = np.array([])
    for index in range(eye_data_sheet.shape[0]):
        start = eye_data_sheet.iloc[index]['start']
        end = eye_data_sheet.iloc[index]['end']
        if start>end:
            start, end = end, start
        #print(start,end)
        raw_data = np.hstack((raw_data, np.linspace(start,end,(end-start)*1000)))

    bout_time_array = group_consecutive(raw_data, step=episode_gap)
    bout_time_pair = [(bout_time_array[index][0], bout_time_array[index][-1]) for index in range(len(bout_time_array))]

    return bout_time_pair

File: juliaLime/analysis.jl
module analysis
using DataFrames
using Distributions

import specialItems: frequency

export horizontal_log_thresh_method

#=  目前计算阈值的方法
    对power的数据取log10, 用标准正态分布函数拟合[参考维基百科正态分布词条],
    得到mu(loc)和sigma(scale).
    阈值取10^(mu+sigma)'=#
function horizontal_log_thresh_method(timeArray, powerArray)
    logData = log10.(powerArray)
    fitModel = fit(Normal,logData)
    mu, sigma = params(fitModel)
    on_thresh = maximum(powerArray) < 0.0005 ? 0.0005 : 10^(mu+sigma)

    dataSheet = DataFrame([timeArray, powerArray, sign.(powerArray .- on_thresh)],[:time, :power, :thresh])
    #distribution_plot(ch_num, dataSheet, on_thresh)

    return dataSheet, on_thresh
end

#=python风格的分割数组
    运算效率感人
    对arr = 260072-element Array{Int64,1}, pos = 88-element Array{Int64,1}
    耗时 0.000545 seconds (136 allocations: 1.996 MiB)
=#
function split!(arr::AbstractArray, pos::AbstractArray)
    #TODO: boundaryCheck

    target = Array{Array}(length(pos)+1)
    target[1] = arr[1:pos[1]]

    for index = 1:length(pos)
        startIndex = pos[index]+1
        endIndex = index == length(pos) ? break : pos[index+1]
        target[index+1] = arr[startIndex:endIndex]
    end
    target[end] = arr[pos[end]+1:end]
    return target
end

#= 寻找datarise的点，并将时间点分割 =#
function group_consecutive(dataSheet; step=4, inname=:thresh, outname=:time)
    pos = find(x->x>0, dataSheet[inname])
    outArray = dataSheet[outname]
    posRisePos = find(x->x>step*frequency, diff(pos))
    posRange = split!(pos, posRisePos)

    boutArray = Array{Array{Float64,1}}(length(posRange))
    for index = 1:length(posRange)
        boutArray[index] = Array{Float64,1}([outArray[posRange[index][1]],outArray[posRange[index][end]]])
    end
    return boutArray
end

end  # module analysis
